### PR TITLE
RPC methods for Validators to Get ShardID, Index and Slot

### PR DESCRIPTION
**Pull request by terenc3t**
Added the following RPC methods for validators to get which **ShardID** it's suppose to be in, which **validator index** of the bitfield and which **slot** it's has to perform for this cycle.

### CODE REVIEW

File: validator/attester/service.go [hunk 1]
```diff
@@ -58,8 +61,9 @@ func NewAttester(ctx context.Context, cfg *Config) *Attester {
 // Start the main routine for an attester.
 func (a *Attester) Start() {
 	log.Info("Starting service")
-	client := a.rpcClientService.AttesterServiceClient()
-	go a.run(a.ctx.Done(), client)
+	attester := a.rpcClientService.AttesterServiceClient()
+	validator := a.rpcClientService.ValidatorServiceClient()
+	go a.run(a.ctx.Done(), attester, validator)
```
Comment (rauljordan) [inline validator/attester/service.go]: no need to pass in a.ctx.Done() here as the routine can access it as a property of the struct

File: beacon-chain/casper/validator.go [hunk 2]
```diff
@@ -156,3 +157,56 @@ func GetProposerIndexAndShard(shardCommittees []*pb.ShardAndCommitteeArray, lcs
...
+		if validators[index].PublicKey == pubKey {
+			return index, nil
+		}
+	}
+
+	return 0, fmt.Errorf("can't find validator index for public key %d", pubKey)
+}
+
+// GetValidatorShardID returns the shard ID of the validator currently participates in.
+func GetValidatorShardID(pubKey uint64, dynasty uint64, validators []*pb.ValidatorRecord, shardCommittees []*pb.ShardAndCommitteeArray) (uint64, error) {
```
Comment (rauljordan) [inline beacon-chain/casper/validator.go]: Prefer to not use `Get` in go function naming. Instead, just use `ValidatorShardID`. Same goes for the other functions.

File: beacon-chain/rpc/service_test.go [hunk 3]
```diff
@@ -47,6 +47,14 @@ func (m *mockChainService) ProcessedAttestationFeed() *event.Feed {
 	return m.attestationFeed
 }
 
+func (m *mockChainService) CurrentCrystallizedState() *types.CrystallizedState {
+	cState, err := types.NewGenesisCrystallizedState()
+	if err != nil {
+		log.Error("Can't get crystallized state")
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service_test.go]: Return nil? What is `cstate` in this branch?

File: shared/bit.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+		chunkLocation++
+	}
+
+	field := bitfield[chunkLocation-1] >> (8 - uint(indexLocation))
+	return field%2 != 0
+}
+
+// BitSetCount counts the number of 1s in a byte using the following algo:
+// https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel
+func BitSetCount(bytes []byte) int {
```
Comment (prestonvanloon) [inline shared/bit.go]: Can we use this library for this? 

https://github.com/steakknife/hamming/blob/16b83346f022ed8fdfc17798a88c44da64b9010d/popcount.go#L60

They have an optimized lookup table which claims to be significantly faster: 

```
BenchmarkCountBitsByte-4            	2000000000	         0.75 ns/op
BenchmarkCountBitsByteAlt-4         	1000000000	         2.37 ns/op
```

`BenchmarkCountBitsByteAlt` is the same implementation as this method, i believe.

File: shared/bit.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	var total int
+	for _, b := range bytes {
+		b = (b & 0x55) + ((b >> 1) & 0x55)
+		b = (b & 0x33) + ((b >> 2) & 0x33)
+		total += int((b + (b >> 4)) & 0xF)
+	}
+	return total
+}
+
+// BitLength returns the length of the bitfield for a giben number of attesters in bytes.
```
Comment (prestonvanloon) [inline shared/bit.go]: Typo: Giben

What does attesters have to do with this method?

File: shared/bit.go (hunk 4 above)
Comment (terenc3t) [reply shared/bit.go]: Good idea. Done!

**Pull request merged**